browser/ai_chat/code_execution_tool.cc

```diff
@@ -0,0 +1,192 @@
+#include "brave/components/ai_chat/core/browser/tools/tool_utils.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "brave/components/script_injector/common/mojom/script_injector.mojom.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_controller.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+#include "third_party/blink/public/mojom/script/script_evaluation_params.mojom.h"
+#include "ui/base/page_transition_types.h"
+#include "url/gurl.h"
+
+namespace ai_chat {
+
+namespace {
+
+constexpr base::TimeDelta kExecutionTimeLimit = base::Seconds(10);
+constexpr char kScriptProperty[] = "script";
+
+std::string WrapScript(const std::string& script) {
+  return base::StrCat({"(async function() { try { ", script,
+                       " } catch (error) { return error.toString(); } })()"});
+}
+
+}  // namespace
+
+CodeExecutionTool::CodeExecutionRequest::CodeExecutionRequest(
+    Profile* profile,
+    const std::string& script,
+    base::TimeDelta execution_time_limit)
+    : content::WebContentsObserver(nullptr), wrapped_js_(WrapScript(script)) {
+  auto otr_profile_id = Profile::OTRProfileID::AIChatCodeExecutionID();
+  auto* otr_profile = profile->GetOffTheRecordProfile(
+      otr_profile_id, /*create_if_needed=*/true);
+  content::WebContents::CreateParams create_params(otr_profile);
+  web_contents_ = content::WebContents::Create(create_params);
+
+  Observe(web_contents_.get());
+
+  web_contents_->GetController().LoadURL(
+      GURL(kAIChatCodeSandboxUIURL), content::Referrer(),
+      ui::PAGE_TRANSITION_TYPED, std::string());
+
+  timeout_timer_.Start(FROM_HERE, execution_time_limit,
+                       base::BindOnce(&CodeExecutionRequest::HandleTimeout,
+                                      base::Unretained(this)));
+}
+
+CodeExecutionTool::CodeExecutionRequest::~CodeExecutionRequest() {
+  Observe(nullptr);
+}
+
+void CodeExecutionTool::CodeExecutionRequest::DidFinishLoad(
+    content::RenderFrameHost* render_frame_host,
+    const GURL& validated_url) {
+  if (!render_frame_host->GetParent() || wrapped_js_.empty()) {
+    return;
+  }
+
+  render_frame_host->GetRemoteAssociatedInterfaces()->GetInterface(&injector_);
+
+  auto wrapped_js_utf16 = base::UTF8ToUTF16(wrapped_js_);
+
+  // Clear the wrapped script to avoid re-using it.
+  wrapped_js_ = {};
+
+  injector_->RequestAsyncExecuteScript(
+      content::ISOLATED_WORLD_ID_GLOBAL, wrapped_js_utf16,
+      blink::mojom::UserActivationOption::kActivate,
+      blink::mojom::PromiseResultOption::kAwait,
```

Comment:
Are we sure we want to call `kAwait` - this means the input script can technically return a promise even though we're asking it to return a string only?